### PR TITLE
replace ONNXOpTransformPass with hybrid transform

### PR DESCRIPTION
**Pull request by sorenlassen**
replaces ONNXOpTransformPass and the repeatOnnxTransform loop in addONNXToMLIRPasses() with hybrid transform, unless disabled with `--onnx-hybrid-pass=false`

set maxNumRewrites in onnx-hybrid-transform to 300 and tested that the model https://huggingface.co/xlnet-large-cased succeeds without warning with this value

prints warning to stderr if onnx-hybrid-transform doesn't converge with the given maxNumRewrites

I tested this PR on the models in the model zoo and the same models passed and failed as with main

possible future work:
- replace const-prop, shape-inference, canonicalizer passes in SimplifyShapeRelatedOpsPass with hybrid transform
- replace some shape-inference/const-prop/canonicalizer passes in addONNXToZHighPasses() with hybrid transform

### CODE REVIEW

**Comment by sorenlassen** [comment]
set maxNumRewrites in onnx-hybrid-transform to 300 and tested that


I tested the model https://huggingface.co/xlnet-large-cased with different maxNumRewrites values and found that it fails with 100:
```
Warning: onnx-hybrid-transform didn't converge with max-num-rewrites=100
Warning: onnx-hybrid-transform didn't converge with max-num-rewrites=100
Assertion failed: (isRankedShapedType(ty) && "Type must be ranked"), function getShape, file TypeUtilities.cpp, line 42.
```
and with 200 it succeeds with one warning:
```
Warning: onnx-hybrid-transform didn't converge with max-num-rewrites=200
```
and with 300 it succeeds wiithout warnings

File: src/Transform/ONNX/ONNXHybridTransformPass.cpp
Comment (tungld) [inline src/Transform/ONNX/ONNXHybridTransformPass.cpp:141]: How did you know that it failed because of "not converge" but other errors like wrong patterns?

Comment (sorenlassen) [reply src/Transform/ONNX/ONNXHybridTransformPass.cpp:141]: The applyPatternsAndFoldGreedily returns false only if there are still patterns to apply when the rewrite limit is reached.

**Comment by chentong319** [comment]
Question to @tungld : is the ConvOpt a really general onnx to onnx transformation that should be included into the hybrid pass? Should it be part of onnx to ZHigh lowering?

**Comment by tungld** [comment]
> Question to @tungld : is the ConvOpt a really general onnx to onnx transformation that should be included into the hybrid pass? Should it be part of onnx to ZHigh lowering?

It's for CPU, and looks like a general transformation. @AlexandreEichenberger may have more comments as he wrote ConvOpt.

File: src/Transform/ONNX/ONNXHybridTransformPass.cpp
Comment (tungld) [reply src/Transform/ONNX/ONNXHybridTransformPass.cpp:141]: Thanks for the explanation! I found `applyPatternsAndFoldGreedily` internally emits the same error message using  LLVM_DEBUG: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp#L731

**Review by tungld — APPROVED** [review]
LGTM. Would be good to have @AlexandreEichenberger commentsOn ConvOpt before merging.

**Comment by AlexandreEichenberger** [comment]
> set maxNumRewrites in onnx-hybrid-transform to 300 and tested that
>I tested the model https://huggingface.co/xlnet-large-cased with different maxNumRewrites values and found that it fails with 100:

Can it be made proportional to the number of ops? That way if we have a very large model in the future, it should scale with it?

**Comment by AlexandreEichenberger** [comment]
Info on the ConvOpt, it transforms a 1x1 Conv into a matmul, as it can be seamlessly implemented as such. I let you decide if it's appropriate for the hybrid transform or not. Its essentially a one-to-one substitution as the two are identical, and the matmul is well optimized whereas the Conv is not yet optimized (for CPU)

**Comment by sorenlassen** [comment]
> > set maxNumRewrites in onnx-hybrid-transform to 300 and tested that
> > I tested the model https://huggingface.co/xlnet-large-cased with different maxNumRewrites values and found that it fails with 100:
> 
> Can it be made proportional to the number of ops? That way if we have a very large model in the future, it should scale with it?

done, it's now calculated as 20 + 0.2 * #ops

@AlexandreEichenberger @tungld I believe I've addressed all feedback and will merge the PR on Monday unless you have more input before then

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM

**Review by chentong319 — APPROVED** [review]
LGTM!

File: src/Transform/ONNX/ONNXHybridTransformPass.cpp
Comment (sorenlassen) [reply src/Transform/ONNX/ONNXHybridTransformPass.cpp:84]: no, but I welcome any ideas

fwiw, now that we agreed to remove ConvOpt we're back to the original vision of grouping the 4 classes of transformations (shape inference, canonicalization, const prop, and decomposition) originally proposed for the hybrid pass 1y ago in PR #1770, so I don't expect that we'll need to add a lot more controls going forward

**Pull request merged**
